Optional rotation handle for selection box overlay + simplify inheritance for Vispy overlays

jacopoabramo opened the pull request:
# References and relevant issues

This PR is a follow-up discussion with @brisvag in reference to a zulip discussion about using overlays for ROI management in image layers (which also triggered issue #7931).

# Description

It adds an optional boolean flag `rotation` in the `InteractionBox` compound to allow the enabling/disabling of the rotation handle. By default this is kept to `True` to maintain backward compatibility with other existing overlays that take advantage of such compound. Some minor changes are applied to other transform box relevant functions to ensure correct behavior.

Additionally, @brisvag identified a problem in the Vispy overlays inheritance to ensure that the correct overlay class is built when updating the `overlay_to_visual` dictionary.

An example is provided in `examples\dev\interaction_box.py`.

The usage of the example is dual fold: it provides a way to demonstrate how to inherit from the selection box overlay to create a custom one, and it uses such custom overlay in conjunction with two mouse event handles to drag and drop the overlay handles on top of the layer to simulate the selection of a sub-region of an image layer.

https://github.com/user-attachments/assets/3724fd42-523b-4fae-82c6-6e09637f8fb2

Furthermore, I took the liberty of making some better type annotations to `InteractionBox`.

jacopoabramo committed:
fix: adjust some type annotations

jacopoabramo committed:
feat: extend interaction box with rotation option and add example

- example currently not working, investigating why

jacopoabramo committed:
fix: inherit directly from `SceneOverlay`

- when inheriting from `SelectionBoxOverlay`, the signals are still bounded to the parent class instead of the child class

jacopoabramo committed:
fix: adjust number of markers

jacopoabramo committed:
Merge branch 'napari:main' into selection-box-rotation

brisvag committed:
simplify inheritance

jacopoabramo committed:
Merge pull request #1 from brisvag/fix/bounding_box_inheritance

Fix inheritance for overlays and simplify code

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

jacopoabramo committed:
Merge branch 'main' into selection-box-rotation

brisvag commented on examples/dev/interaction_box.py:
```diff
@@ -0,0 +1,73 @@
+                bot_right[::-1],
+                handles=self.overlay.handles,
+                selected=self.overlay.selected_handle,
+                rotation=False,
```
Worth adding a comment here to show that this line is the only change from the superclass.

brisvag commented on examples/dev/interaction_box.py:
```diff
@@ -0,0 +1,73 @@
+# define a model for the selection box overlay;
+# we subclass from SelectionBoxOverlay to get the
+# default behavior of the selection box; we redefine
+# the handles attribute to True to show the handles
```
this comment is outdated I think?

brisvag commented on examples/dev/interaction_box.py:
```diff
@@ -0,0 +1,73 @@
+
+image._overlays['selection_no_rotation'] = SelectionBoxNoRotation(
+    bounds=[(0, 0), data.shape], handles=True
+)
```
note that this bounds will look wrong on the canvas, because everything is offset by 0.5. This is because image data is located with the first pixel at the origin. You might wanna change this to account for that. You should be able to use `image._display_bounding_box_augmented((0, 1)).T` to get the right values.

brisvag commented on examples/dev/interaction_box.py:
```diff
@@ -0,0 +1,73 @@
+        warnings.simplefilter('ignore')
+        transform = viewer.layers['image']._overlays["transform_box"]
+        selection = viewer.layers['image']._overlays['selection_no_rotation']
+        transform.visible = transform_box
```
I would remove the transform box from here. Note that you can show it by switching to transform mode from the top left buttons.

brisvag reviewed (approved):
A few comments, but looks great in general!

jacopoabramo committed:
fix: add more meaningful comments and simplify toggle_overlay

- we toggle only the new overlay; the transform box has its own dedicated mode button

jacopoabramo committed:
fix: use correct layer bounds

jacopoabramo committed:
Merge branch 'selection-box-rotation' of https://github.com/jacopoabramo/napari into selection-box-rotation

jacopoabramo commented on examples/dev/interaction_box.py:
```diff
@@ -0,0 +1,85 @@
+# we recover the bounds of the image layer;
+# this method will ensure that the overlay is drawn
+# correctly in the viewer
+layer_bounds = tuple(tuple(x.tolist()) for x in image._display_bounding_box_augmented([0, 1]).T)
```
I must admit I'm not a great fan of this because it feels a bit too convoluted to recover the bounds. Probably a dedicated `visual_bounds` property could be provided in the future?

brisvag replied:
I'm not sure why the validator doesn't handle converting array to tuples... but otherwise this is as intended. In general, the "visual bounds" are not a thing that the user should care about, unlike the data extents.

In fact, this stuff might be better moved to the vispy module in the future, as this is vispy specific and might differ with other backends.

jacopoabramo committed:
fix: rework overlay example to handle mouse callbacks

- example is currently broken, will be fixed asap

jacopoabramo committed:
Merge branch 'main' into selection-box-rotation

jacopoabramo committed:
Merge branch 'napari:main' into selection-box-rotation

jacopoabramo committed:
fix: hide the rotation handle edge if marker is disabled

jacopoabramo committed:
fix: correctly calculate the bounds based on the new mouse position

jacopoabramo committed:
Merge branch 'selection-box-rotation' of https://github.com/jacopoabramo/napari into selection-box-rotation

jacopoabramo committed:
fix: cast directly to MousePosition

jacopoabramo committed:
fix: fix comment

brisvag committed:
some fixes

jacopoabramo committed:
Merge pull request #2 from brisvag/fix/bounding_box_mouse

some fixes

jacopoabramo committed:
Merge branch 'napari:main' into selection-box-rotation

brisvag committed:
use mro to get closes superclass

TimMonko committed:
type layer_to_visual

jacopoabramo committed:
Merge branch 'napari:main' into selection-box-rotation

jacopoabramo committed:
fix: add *args and **kwargs for typing coverage

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

brisvag reviewed (approved):
I think we can merge this!

TimMonko reviewed (approved):
Thank you Jacopo! Glad we got this there. Great work! Looking forward to the next one :wink:
